### ISSUE TITLE
Silent failure to save files on network drive when folder mode has been used on a large directory (0.9.48)

### ISSUE THREAD

**Issue by pjeby** [typora/typora-issues#1445]
At some point today, my copy of Typora started silently failing to save edited files.  I *think* it was after launching Typora from the command line on a non-existent filename, but it's possible it happened before that point and I overlooked it.

The specific behavior is that saving to a network drive, Typora acts like it saves, but doesn't do anything.  The first indication there's a problem (other than the file contents not changing) is when exiting: it displays a warning that the file needs to be saved (and the file status goes back to modified).  Choosing to save acts again as though the file is saved, but once again does not save.

The only way to get Typora to *actually* save the file is to use Save As... and manually overwrite the same file.  The fact that Save As... works seems to indicate that this is an issue with Typora, not the network drive.  (Editing the same file(s) in other editors, the contents save just fine, but Typora does not notice that the other editors have changed the file and does not reload them.)

Inspecting typora.log, there are no messages to indicate anything wrong, just:

```
INFO 4/26/2018, 1:58:01 AM  ------------------start------------------ 
INFO 4/26/2018, 1:58:02 AM  addRecentDocument
INFO 4/26/2018, 1:58:02 AM  C:\Program Files\Typora\resources\app\locales\Base.lproj\Menu.json
INFO 4/26/2018, 2:00:29 AM  [ 'C:\\Program Files\\Typora\\Typora.exe',
  'S:\\home\\pje\\bedrock\\post.md' ]
INFO 4/26/2018, 2:00:29 AM  addRecentDocument
INFO 4/26/2018, 2:00:30 AM  secondary instance would exit
INFO 4/26/2018, 2:02:34 AM  [ 'C:\\Program Files\\Typora\\Typora.exe',
  'S:\\home\\pje\\bedrock\\post.md' ]
INFO 4/26/2018, 2:02:34 AM  addRecentDocument
INFO 4/26/2018, 2:02:35 AM  secondary instance would exit
INFO 4/26/2018, 2:05:56 AM  ------------------quiting------------------
```

**Comment by pjeby** [comment]
Just to make things more interesting...  A few minutes ago, one of my other editors detected a change in a file I still had open in Typora, as if Typora were saving the file, but doing so several minutes *after* I told it to.  I suspect now that the issue is that Typora is showing in the UI that it has saved the file, even though the actual save process hasn't yet begun, let alone succeeded.

(It would probably be helpful if there was some visible indication that the save hasn't yet succeeded, such as a cancellable progress dialog.)

I should also note that no backups are being made of the not-saved file contents.  Presumably it's because Typora believes it *is* saving, even though it has not actually done so yet.

**Comment by pjeby** [comment]
Things that have *not* fixed the problem so far:

* Downgrading to 0.9.47
* Rebooting the local Windows machine
* Reinstalling 0.9.48
* Restarting the remote Samba server
* Trying a different remote Samba server

What I *have* found, though, is that the issue appears to be related to how many files or subdirectories are underneath the directory where the file is located, and whether folder view has ever been in effect for that window.  The file I've been having trouble with resides in the root directory of a project with nearly 25,000 files in various directories underneath it.

If I move the file to a subdirectory without any other files, then there is an initial delay during which saving doesn't work, but a minute or so later, saving starts working.

If I open Typora in outline mode, the problem doesn't happen: saving works immediately.  If I then switch to the folder view, the problem manifests.  Switching back to outline view does not fix the problem; I have to close the window and re-open it to get it to work again.

So it seems it is a problem with the folder view interacting with saves when the folder has lots of files under it on a remote connection (thus making it take a long time to refresh?).

**Comment by abnerlee** [comment]
> I think it was after launching Typora from the command line on a non-existent filename

This should pop up a file open fail dialog, so you won't see the error dialog for this?

---

And, could you provide log from console when save fail happens, you could grab it from `View` -> `Toggle DevTools` from menubar?

**Comment by pjeby** [comment]
There was nothing in the log, because the actual problem appears to be that the saving is *delayed* rather than failed.  The delay only happens if the sidebar is displaying files, and then only if the displayed directory has lots of files under it (even under other subdirectories), on a network drive.  (This implies that there is always a delay related to that, it's just small when done on the local HD.)

So, it seems that saving is done in the background, but for some reason waits on the folder information to refresh (or forces it to?) before actually saving the file.  But in the UI it looks like the save is complete, even though it hasn't started yet.  Running a inotify trace on the directory indicates that Typora doesn't do anything to the file until later -- sometimes several minutes later, after I've already switched to a different file in the folder view and told it to discard the (then-current) changes in the original file.

For example, right now, while typing this, I saw my inotify trace show that Typora *just now* wrote to a file that I saved about 3 minutes ago.  In the intervening 3 minutes, the file contents were the same on disk as they were before I changed them, but now they match what I told it to save 3 minutes ago.

So, there is no actual failure to save.  Instead, Typora's UI *shows* that a save has succeeded, but in reality the file's contents will not be rewritten until much later.  In the interim, changing between files or exiting will pop up a save/discard warning, indicating that at least *that* part of Typora knows the save isn't finished.

Anyway, there are two actual problems here:

1. The state of the UI does not reflect the ongoing nature of the save process, leaving one to believe the file is saved when it is not (and is not in the recovery bin, either).

2. The save is being deferred until after some lengthy process related in some way to the sidebar.  I don't typically use the "article" view, so I don't know if it applies to that view as well.

The combination of these problems makes it look like there is a failure to save, but in reality it is just deferred while looking like it's not.  In practical terms it is hugely challenging because you can't get a file's contents to update for several minutes, and if you make any changes during those minutes then you must once again wait to save *them*, and so on.

At the present time my workaround is simply to not use the file sidebar, and if I accidentally open it, to close the window and re-open a new one.

**Comment by abnerlee** [comment]
@pjeby We fix a saving bug caused by race condition, please check if the new updates can fix this issue or not.

**Comment by pjeby** [comment]
No, it's not fixed; the writes are still delayed, not showing up until minutes later if the folder view is used at all.  But if the same file is opened with outline view and folder view is never used, then the writes happen immediately.  This pattern is consistently reproducible for the target directory, which contains around 25,000 files in dozens of deeply nested subdirectories and is being accessed over a remote VPN connection.  It's likely that the delay is only noticeable cases of high data volume and low transfer speed.

**Comment by seiks** [comment]
Hi.

I am having a related issue and thought I'd chip in on the discussion (even though this issue is old). This on version **0.9.9.20.3**.

The problem I'm facing is that saving on a NTFS-formatted SMB share fails.

<img width="426" alt="0" src="https://user-images.githubusercontent.com/22685871/51190708-82105380-18eb-11e9-9449-8f1e0d2259e8.png">

When I look into the folder where the open file is, I see that Typora has created some temporary files which are locked (cannot be removed).

<img width="834" alt="1" src="https://user-images.githubusercontent.com/22685871/51190713-85a3da80-18eb-11e9-90c1-ff0fa8a04091.png">

Briefly, there also appears a folder "(A Document Being Saved By Typora)". I assume Typora is doing some form of atomic saving, and for some reason it fails on a network share.

Do @abnerlee have any idea what is causing this?

**Comment by abnerlee** [comment]
@seiks The version control is managed by macOS system. So could you also reproduce this using TextEdit.app ?

**Comment by seiks** [comment]
@abnerlee You are right on the version control. I can confirm that last warning appears also on TextEdit, so it is not a Typora-related issue.

However, the saving problem still persists, although I managed to isolate it a little bit. It appears that if you press CMD+S (Save) when there are no changes to the file, Typora breaks and starts throwing errors. In the GIF below, Typora breaks exactly at the moment I press save when there are no changes, however you can see for yourself that the couple of initial saves before this work just fine.

![2019-01-15 20 49 19](https://user-images.githubusercontent.com/22685871/51202485-80548900-1907-11e9-941d-1c8c4bc8e7a8.gif)

For reference, here's a GIF of me saving a file in TextEdit. I also tried the above (saving when no changes present) on TextEdit, but TextEdit didn't throw any errors about it, so I think that there is some form of bug in Typora.

![2019-01-15 20 48 23](https://user-images.githubusercontent.com/22685871/51202480-7c286b80-1907-11e9-8f05-005c8162c034.gif)

**Comment by seiks** [comment]
@abnerlee You are right on the version control. I was actually able to reproduce these same issues on TextEdit, so I guess Typora isn't the culprit here.

Sorry for jumping the gun on you.

**Comment by abnerlee** [comment]
put into #3981
